File: src/ipv4.cpp
#include "mini_traceroute/ipv4.hpp"

#include <string>

namespace mtr {

std::optional<Ipv4Header> parse_ipv4_header(const std::uint8_t* data, std::size_t len) {
  if (data == nullptr || len < 20) return std::nullopt;

  const std::uint8_t version = static_cast<std::uint8_t>(data[0] >> 4);
  const std::uint8_t ihl = static_cast<std::uint8_t>(data[0] & 0x0F);
  if (version != 4 || ihl < 5) return std::nullopt;

  const std::size_t header_bytes = static_cast<std::size_t>(ihl) * 4;
  if (header_bytes > len) return std::nullopt;

  Ipv4Header h;
  h.ihl_words = ihl;
  h.total_length = static_cast<std::uint16_t>((data[2] << 8) | data[3]);
  h.protocol = data[9];
  h.src_addr = (static_cast<std::uint32_t>(data[12]) << 24) |
               (static_cast<std::uint32_t>(data[13]) << 16) |
               (static_cast<std::uint32_t>(data[14]) << 8) |
               static_cast<std::uint32_t>(data[15]);
  h.dst_addr = (static_cast<std::uint32_t>(data[16]) << 24) |
               (static_cast<std::uint32_t>(data[17]) << 16) |
               (static_cast<std::uint32_t>(data[18]) << 8) |
               static_cast<std::uint32_t>(data[19]);
  return h;
}

std::string ipv4_to_string(std::uint32_t addr_host_order) {
  return std::to_string((addr_host_order >> 24) & 0xFF) + "." +
         std::to_string((addr_host_order >> 16) & 0xFF) + "." +
         std::to_string((addr_host_order >> 8) & 0xFF) + "." +
         std::to_string(addr_host_order & 0xFF);
}

}  // namespace mtr

File: docs/assets/ipv4.js
// Minimal IPv4 header view — a port of src/ipv4.cpp. The C++ core defines its own wire
// struct instead of including <netinet/ip.h>; this file is the same idea in JavaScript, so
// the page parses the bytes it displays rather than trusting a value carried alongside them.

export const PROTOCOL_ICMP = 1;
export const PROTOCOL_UDP = 17;

/**
 * Parse an IPv4 header from the front of `data` (optionally at an offset).
 * @returns {{ihlWords:number, protocol:number, totalLength:number, ttl:number,
 *            srcAddr:number, dstAddr:number, headerBytes:number} | null}
 *          null when the buffer is too short, the version is not 4, or the declared
 *          header length is invalid.
 */
export function parseIpv4Header(data, offset = 0, len = data.length - offset) {
  if (!data || len < 20) return null;

  const version = data[offset] >> 4;
  const ihl = data[offset] & 0x0f;
  if (version !== 4 || ihl < 5) return null;

  const headerBytes = ihl * 4;
  if (headerBytes > len) return null;

  return {
    ihlWords: ihl,
    totalLength: (data[offset + 2] << 8) | data[offset + 3],
    ttl: data[offset + 8],
    protocol: data[offset + 9],
    srcAddr: readAddr(data, offset + 12),
    dstAddr: readAddr(data, offset + 16),
    headerBytes,
  };
}

/** Format a host-order address as dotted decimal (0x01020304 -> "1.2.3.4"). */
export function ipv4ToString(addrHostOrder) {
  return [24, 16, 8, 0].map((shift) => (addrHostOrder >>> shift) & 0xff).join('.');
}

/** Parse dotted decimal into a host-order 32-bit address. */
export function ipv4FromString(dotted) {
  const parts = dotted.split('.').map(Number);
  if (parts.length !== 4 || parts.some((n) => !Number.isInteger(n) || n < 0 || n > 255)) {
    throw new Error(`not a dotted IPv4 address: ${dotted}`);
  }
  return ((parts[0] << 24) | (parts[1] << 16) | (parts[2] << 8) | parts[3]) >>> 0;
}

function readAddr(data, at) {
  return ((data[at] << 24) | (data[at + 1] << 16) | (data[at + 2] << 8) | data[at + 3]) >>> 0;
}
